File: backend_go/models/stock_movement.go
package models

import (
	"time"
)

type StockMovementType string

const (
	Initial    StockMovementType = "initial"
	Sale       StockMovementType = "sale"
	Restock    StockMovementType = "restock"
	Return     StockMovementType = "return"
	Adjustment StockMovementType = "adjustment"
)

type StockMovement struct {
	ID          uint `gorm:"primaryKey"`
	OrderID     *uint
	ProductID   uint              `gorm:"column:product_id"`
	Product     Product           `gorm:"foreignKey:ProductID"`
	Type        StockMovementType `gorm:"not null"`
	Quantity    int               `gorm:"not null"`
	CreatedAt   time.Time         `gorm:"not null"`
	Description string
}

type StockMovementResponse struct {
	ID          uint              `json:"id"`
	ProductID   uint              `json:"product_id"`
	ProductName string            `json:"product_name"`
	Type        StockMovementType `json:"type"`
	Quantity    int               `json:"quantity"`
	CreatedAt   time.Time         `json:"created_at"`
	Description string            `json:"description"`
	OrderID     *uint             `json:"order_id"`
}

File: shared_dtos/src/stock_movement.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[cfg_attr(feature = "sqlx", derive(sqlx::Type))]
#[cfg_attr(feature = "sqlx", sqlx(type_name = "TEXT", rename_all = "snake_case"))]
#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(feature = "ts", ts(export, export_to = "stock_movement.ts"))]
#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StockMovementType {
    Initial,
    Restock,
    Sale,
    Return,
    Adjustment,
}

#[cfg_attr(feature = "openapi", derive(utoipa::ToSchema))]
#[cfg_attr(feature = "ts", derive(ts_rs::TS))]
#[cfg_attr(
    feature = "ts",
    ts(export, export_to = "stock_movement.ts", rename = "StockMovement")
)]
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StockMovementAdminResponse {
    pub id: i64,
    pub order_id: Option<i64>,
    pub product_id: i64,
    pub product_name: String,
    pub r#type: StockMovementType,
    pub quantity: i64,
    pub created_at: DateTime<Utc>,
    pub created_by: i64,
    pub description: Option<String>,
}
